### PR TITLE
[1LP][RFR] Storage Manager [Block and Object] WT

### PR DESCRIPTION
**Pull request by digitronik**
Purpose or Intent
=================
- added combine navigation destination for object and block storage manager.
- added simple navigation test for block and object storage manager.

{{pytest: cfme/tests/storage/test_manager.py -v --use-provider rhos7-ga}}

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit cd21cbec3007d82d0a66b86c3bc8ef8312b2fc3e:

`cfme/storage/manager.py`:
- :bangbang: Line 193:45: **E127** *continuation line over-indented for visual indent*
- :bangbang: Line 193:69: **E231** *missing whitespace after ','*
- :bangbang: Line 215:69: **E231** *missing whitespace after ','*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/storage/manager.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+                                  PaginationPane,
+                                  SummaryTable,
+                                  Table
+                                  )
+from widgetastic_patternfly import (Button,
+                                    Dropdown,
+                                    FlashMessages
+                                    )
+from widgetastic.widget import View, NoSuchElementException, Text
+from cfme.exceptions import ItemNotFound
```
Comment (quarckster) [inline cfme/storage/manager.py]: **REQUIRED**
Please follow style guide for imports.

File: cfme/storage/manager.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+        return(
+            self.logged_in_as_current_user and
+            self.navigation.currently_selected == nav)
+
+
+class ManagerAllView(ManagerView):
+    """The all Storage Manager or Provider page"""
+    @property
+    def is_displayed(self):
+        """Is this page currently being displayed"""
```
Comment (quarckster) [inline cfme/storage/manager.py]: Here and below a docstring for `is_displayed` property not needed.

File: cfme/tests/storage/test_manager.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+pytestmark = [pytest.mark.tier(3),
+              test_requirements.storage,
+              pytest.mark.usefixtures('openstack_provider', 'setup_provider')]
+
+
+@pytest.mark.uncollectif(lambda provider: not provider.one_of(OpenStackProvider))
+def test_block_manager_navigation(openstack_provider, appliance):
+    """Simple test for navigation destinations"""
+    block_name = openstack_provider.name + ' Cinder Manager'
```
Comment (quarckster) [inline cfme/tests/storage/test_manager.py]: **REQUIRED**
Please use `format`

**Comment by dajoRH** [comment]
Lint report for commit 80f662c0121307c398dffc20e1d6739d597728f7:

`cfme/storage/manager.py`:
- :bangbang: Line 15:5: **E123** *closing bracket does not match indentation of opening bracket's line*
- :bangbang: Line 20:5: **E123** *closing bracket does not match indentation of opening bracket's line*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by digitronik** [comment]
@quarckster added required changes.

**Comment by digitronik** [comment]
@quarckster collection added. Due to the impact of the collection, i removed the excessive block and object manages classes. The required attributes added in respective collection classes only.

File: cfme/storage/manager.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    cancel_button = Button("Cancel")
+
+    @property
+    def is_displayed(self):
+        return (
+            self.in_manager and
+            self.breadcrumb.active_location == "'Storage Manager' Policy Assignment")
+
+
+class BlockManagerCollection(NavigatableMixin):
```
Comment (quarckster) [inline cfme/storage/manager.py]: **REQUIRED**
Please read today's last email from @psav http://cfme-tests.readthedocs.io/guides/collections.html

**Comment by digitronik** [comment]
@quarckster  improved with a new collection policy, remove navMixin and added BaseCollection and BaseEntity with other changes.

File: cfme/tests/storage/test_manager.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+pytestmark = [pytest.mark.tier(3),
+              test_requirements.storage,
+              pytest.mark.usefixtures('openstack_provider', 'setup_provider')]
+
+
+@pytest.mark.uncollectif(lambda provider: not provider.one_of(OpenStackProvider))
+def test_block_manager_navigation(openstack_provider, appliance):
+    """Simple test for navigation destinations"""
+
+    collection = BlockManagerCollection(appliance=appliance)
```
Comment (quarckster) [inline cfme/tests/storage/test_manager.py]: **REQURIED**
This should be in a fixture.

File: cfme/tests/storage/test_manager.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    assert view.is_displayed
+
+    view = navigate_to(block_manager, 'Details')
+    assert view.is_displayed
+
+    block_manager.refresh()
+
+
+@pytest.mark.uncollectif(lambda provider: not provider.one_of(OpenStackProvider))
+def test_object_manager_navigation(openstack_provider, appliance):
```
Comment (quarckster) [inline cfme/tests/storage/test_manager.py]: **REQUIRED**
These tests are identical, so it's better to have only one with the name let's say ` test_managers_navigation` and parametrized fixture.

Comment (digitronik) [reply cfme/tests/storage/test_manager.py]: @quarckster  added.

File: cfme/storage/manager.py [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+    policy = Dropdown('Policy')
+
+
+class ManagerDetailsToolbar(View):
+    """The toolbar on the Storage Manager or Provider detail page"""
+    reload = Button(title='Reload Current Display')
+    configuration = Dropdown('Configuration')
+    policy = Dropdown('Policy')
+    monitoring = Dropdown('Monitoring')
+    download = Button(title='Download summary in PDF format')
```
Comment (izapolsk) [inline cfme/storage/manager.py]: QUESTION: are you sure it always has such title ? 
Besides that, are you sure this is Button not Dropdown ?

File: cfme/storage/manager.py [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+    """A base view for all the Storage Manager or Provider pages"""
+    title = Text('.//div[@id="center_div" or @id="main-content"]//h1')
+
+    flash = FlashMessages(
+        './/div[@id="flash_msg_div"]/div[@id="flash_text_div" or '
+        'contains(@class, "flash_text_div")]')
+
+    @property
+    def in_manager(self):
+        nav = self.context['object'].nav.pick(self.context['object'].appliance.version)
```
Comment (izapolsk) [inline cfme/storage/manager.py]: REQUIRED: this is good idea to move navigation path to class but  I don't like this name - "nav". It doesn't explain what it contains even more it sounds similar to navmazing object/property. I'd advice to call it nav_path or navigation_path for example.

File: cfme/storage/manager.py [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+
+
+class ManagerTagsView(ManagerView):
+    """The tag page for Storage Manager or Provider"""
+    breadcrumb = BreadCrumb()
+    select_tag = BootstrapSelect('tag_cat')
+    select_value = BootstrapSelect('tag_add')
+    save_button = Button('Save')
+    reset_button = Button('Reset')
+    cancel = Button('Cancel')
```
Comment (izapolsk) [inline cfme/storage/manager.py]: Please either add _button ending everywhere or don't use it at all

File: cfme/storage/manager.py [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+        self.nav = VersionPick({
+            Version.lowest(): ['Storage', 'Storage Providers'],
+            '5.8': ['Storage', 'Object Storage', 'Managers']})
+        self.type = 'Object Storage Managers'
+
+    def instantiate(self, name, provider):
+        return Manager(self, name, provider)
+
+
+class Manager(BaseEntity):
```
Comment (izapolsk) [inline cfme/storage/manager.py]: REQUIRED: Please use full name. class name, etc should always be self explaining. it makes code easily readable.

File: cfme/storage/manager.py [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+    """ Model of an storage manager in cfme
+
+    Args:
+        name: Name of the object manager.
+        provider: provider
+        appliance: appliance
```
Comment (izapolsk) [inline cfme/storage/manager.py]: REQUIRED: there is no such argument now

File: cfme/storage/manager.py [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+    """
+
+    def __init__(self, collection, name, provider):
+        self.collection = collection
+        self.appliance = self.collection.appliance
+        self.name = name
+        self.provider = provider
+        self.nav = self.collection.nav
+        self.type = self.collection.type
+        self.name_vari = VersionPick({Version.lowest(): 'Storage Provider',
```
Comment (izapolsk) [inline cfme/storage/manager.py]: REQUIRED: please give readable name to variable

File: cfme/storage/manager.py [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+                                  '5.8': 'Storage Manager'}).pick(self.appliance.version)
+
+    def refresh(self, cancel=False):
+        """Refresh storage manager"""
+        view = navigate_to(self, 'Details')
+        view.toolbar.configuration.item_select('Refresh Relationships and Power States',
+                                               handle_alert=not cancel)
+
+        if not cancel:
+            msg = "Refresh Provider initiated for 1 {} from the CFME Database"\
```
Comment (izapolsk) [inline cfme/storage/manager.py]: REQUIRED: please get rid of back slash

File: cfme/tests/storage/test_manager.py [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+@pytest.mark.uncollectif(lambda provider: not provider.one_of(OpenStackProvider))
+def test_manager_navigation(collection_manager):
+    collection, manager = collection_manager
+    view = navigate_to(collection, 'All')
+    assert view.is_displayed
+
+    view = navigate_to(manager, 'Details')
+    assert view.is_displayed
+
+    manager.refresh()
```
Comment (izapolsk) [inline cfme/tests/storage/test_manager.py]: Q: what is it for ?

File: cfme/storage/manager.py (hunk 7 above)
Comment (digitronik) [reply cfme/storage/manager.py]: yes, it's not a dropdown, a simple button with pdf icon, having title "Download summary in PDF format"

File: cfme/tests/storage/test_manager.py (hunk 14 above)
Comment (digitronik) [reply cfme/tests/storage/test_manager.py]: @izapolsk, Its completely new page so added simple navigation test. while navigating added refresh method to capture its flash [Idea came when I added one BZ for the same refresh from 'All' Page]. this test going to remove as I start automating tests for this page.

**Comment by digitronik** [comment]
@izapolsk changes committed. Please review as you free.

File: cfme/storage/manager.py [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+        return(
+            self.title.text == expected_title and
+            self.entities.breadcrumb.active_location == expected_title)
+
+    toolbar = View.nested(ManagerDetailsToolbar)
+    sidebar = View.nested(ManagerDetailsAccordion)
+    entities = View.nested(ManagerDetailsEntities)
+
+
+class ManagerTagsView(ManagerView):
```
Comment (izapolsk) [inline cfme/storage/manager.py]: @psav , @rlbabyuk, tags/policies views are the same everywhere. Do we have any such shared views to re-use here ?

File: cfme/storage/manager.py [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+        return (
+            self.in_manager and
+            self.breadcrumb.active_location == 'Tag Assignment')
+
+
+class ManagePoliciesView(ManagerView):
+    """The policies page for Storage Manager or Provider"""
+    breadcrumb = BreadCrumb()
+    policies = BootstrapTreeview("protectbox")
+    save_button = Button("Save")
```
Comment (izapolsk) [inline cfme/storage/manager.py]: REQUIRED: as I already wrote before please either use _button everywhere or remove it everywhere.

File: cfme/storage/manager.py [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+            self.in_manager and
+            self.breadcrumb.active_location == "'Storage Manager' Policy Assignment")
+
+
+class BlockManagerCollection(BaseCollection):
+    """Collection object [block manager] for the :py:class:'cfme.storage.manager'"""
+
+    def __init__(self, appliance):
+        self.appliance = appliance
+        self.nav = VersionPick({
```
Comment (izapolsk) [inline cfme/storage/manager.py]: REQUIRED: as I already wrote before, please use another name

File: cfme/storage/manager.py [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+        name: Name of the object manager.
+        provider: Provider
+    """
+
+    def __init__(self, collection, name, provider):
+        self.collection = collection
+        self.appliance = self.collection.appliance
+        self.name = name
+        self.provider = provider
+        self.nav = self.collection.nav
```
Comment (izapolsk) [inline cfme/storage/manager.py]: ^^

File: cfme/storage/manager.py [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+        self.nav = VersionPick({
+            Version.lowest(): ['Storage', 'Storage Providers'],
+            '5.8': ['Storage', 'Object Storage', 'Managers']})
+        self.manager_type = 'Object Storage Managers'
+
+    def instantiate(self, name, provider):
+        return BaseManager(self, name, provider)
+
+
+class BaseManager(BaseEntity):
```
Comment (izapolsk) [inline cfme/storage/manager.py]: REQUIRED: can you please use some self explaining name here ? for example StorageManager, etc.

If you run into BaseManager class somewhere in another file, you won't be able to get what is that.
So, class names or public properties should have self explaining names.

File: cfme/storage/manager.py (hunk 15 above)
Comment (digitronik) [reply cfme/storage/manager.py]: @izapolsk   Yup you are right, I referred @rlbabyuk  PR about TagPageView and WidgetasticTaggable. removing it.

File: cfme/storage/manager.py (hunk 18 above)
Comment (digitronik) [reply cfme/storage/manager.py]: @izapolsk, I have to take nav [navigation_path] attribute from the collection. it's helping me to suppress "super method" with two different classes.

Comment (izapolsk) [reply cfme/storage/manager.py]: I already wrote you about nav name. it is confusing and doesn't explain what it contains.

**Comment by dajoRH** [comment]
Lint report for commit c1d659f6d158177393c374f28b084347759ab966:

`cfme/storage/manager.py`:
- :bangbang: Line 79:65: **E126** *continuation line over-indented for hanging indent*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by digitronik** [comment]
@izapolsk , changes
nav > naviagation_path
Manager > StorageManager, as well as all Views, name also changed.
added, WidgetasticTaggeble method to avoid tag view.

Ready for Review

**Comment by izapolsk** [comment]
super, now test name is much more readable

**Pull request merged**